src/sim.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+import math
+import numpy as np
+import pandas as pd
+import salabim as sim
+import sys
+
+
+#  based on SimPy example model
+class Bus(sim.Component):
+    """
+    TODO
+    """
+    def __init__(self, start_time, end_time, trip_power, charging_rate, *args, **kwargs):
+        sim.Component.__init__(self, *args, **kwargs)
+        self.start_time    = start_time
+        self.end_time      = end_time
+        self.trip_power    = trip_power
+        self.charging_rate = charging_rate
+
+    def process(self):
+        """TODO
+        """
+        #Trip starts when this component is activated
+        yield self.request(bus_depot)
+        yield self.hold(till=self.end_time)
+        yield self.request((energy, self.trip_power))
+        yield self.hold(duration=self.trip_power/self.charging_rate*3600)
+        self.release(bus_depot)
+
+
+
+class Dispatcher:
+    def __init__(self, input_file, battery_cap_kwh=100, kwh_per_km=2, charging_rate=150):
+        self.gtfs            = pd.read_pickle(input_file, compression='infer')
+        self.gtfs            = self.gtfs.sort_values(['block_id', 'start_arrival_time'])
+        self.gtfs            = self.gtfs[~self.gtfs.trip_id.str.contains('RAIL')]
+        self.buses           = []
+        self.battery_cap_kwh = battery_cap_kwh
+        self.kwh_per_km      = kwh_per_km
+        self.charging_rate   = charging_rate
+        self.trip_seg_count  = 0
+        for block_id, group in self.gtfs.groupby(['block_id']):
+            self.schedule_block(block_id, group)
+
+    def schedule_block(self, block_id, trips):
+        trips = trips.copy()
```

Comment:
why does it need its own copy? 

should this be: 

```suggestion
self.trips = trips
```
instead?

Reply:
Because a column is added directly below this, but only to the group. That behaviour raises warning flags. Columns should only be added to the entire data frame.

Reply:
ah, makes sense.